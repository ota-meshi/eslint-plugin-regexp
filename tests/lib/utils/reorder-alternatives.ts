// eslint-disable-next-line eslint-comments/disable-enable-pair -- x
/* eslint-disable func-style, no-restricted-imports -- x */
import chai from "chai"
import { RegExpParser, visitRegExpAST } from "regexpp"
import type { Alternative, Node } from "regexpp/ast"
import type { OptionalMatchingDirection } from "../../../lib/utils/reorder-alternatives"
import {
    canReorder,
    getDeterminismEqClasses,
} from "../../../lib/utils/reorder-alternatives"
import type { RegExpContext, ToCharSet } from "../../../lib/utils"
import * as RAA from "regexp-ast-analysis"
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot"

chai.use(jestSnapshotPlugin())

describe(canReorder.name, function () {
    describe("reorder all", function () {
        const valid: RegExp[] = [
            /0|1|2|3/,
            /0|0|1|1|2|3|44/,
            /\b(a|b|aa|\w|\d)\b/,
            /foo|bar/,
            /(int|integer)\b/,
        ]
        const invalid: RegExp[] = [/int|integer/]

        function canReorderAll(regex: RegExp): boolean {
            const context = getFakeRegExpContext(regex)

            let all = true

            visitParents(context.patternAst, (parent) => {
                if (!canReorder(parent.alternatives, context)) {
                    all = false
                }
            })

            return all
        }

        for (const regex of valid) {
            it(regex.toString(), function () {
                chai.assert.equal(canReorderAll(regex), true)
            })
        }

        for (const regex of invalid) {
            it(regex.toString(), function () {
                chai.assert.equal(canReorderAll(regex), false)
            })
        }
    })
})

describe(getDeterminismEqClasses.name, function () {
    const directionIndependentRegexes: RegExp[] = [
        /abc/,
        /a|b|c|d/,
        /a|aa|aaa|b|bb|bbb/,
        /a|b|c|d|[a-c]/,
        /ab|bc|ca/,
        /\bcircle|ellipse|closest|farthest|contain|cover\b/,
        /\p{L}|[a-z]|\d+/u,
        /(int|integer)\b/,
        /device_ios_(?:ipad|ipad_retina|iphone|iphone5|iphone6|iphone6plus|iphone_retina|unknown)\b/,

        // this is an interesting example because a naive algorithm might
        // take 2^20 steps
        /aaaaaaaaaaaaaaaaaaaa|bbbbbbbbbbbbbbbbbbbb|[^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^]/,
        /a{20}|b{20}|[^]{20}/,
        /a{20}c|b{20}a|[^]{20}b/,
        /(?:script|source)_foo|sample/,
    ]
    const directionalRegexes: RegExp[] = [
        /abc|a/,
        /aa\d+|ba\d+/,
        /a.*b|b.*b|c.*a/,
        /a.*b|b.*b|b.*a/,
        /int|integer/,
        /a{20}c|b{20}a|[^]{19}b/,

        /anchor_1_x|anchor_1_y|anchor_2_x|anchor_2_y|reaction_force_x|reaction_force_y|reaction_torque|motor_speed|angle|motor_torque|max_motor_torque|translation|speed|motor_force|max_motor_force|length_1|length_2|damping_ratio|frequency|lower_angle_limit|upper_angle_limit|angle_limits|max_length|max_torque|max_force/,
        />>=?|<<=?|->|--|\+\+|&&|\|\||[?:~]|<=>|[-+*/%&|^!=<>]=?|\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\b/,
    ]

    type AllEqClasses = Record<
        OptionalMatchingDirection,
        Record<string, string[]>
    >

    function getAllEqClasses(regex: RegExp): AllEqClasses {
        const directions: OptionalMatchingDirection[] = [
            "ltr",
            "rtl",
            "unknown",
        ]
        const result: AllEqClasses = { ltr: {}, rtl: {}, unknown: {} }

        const context = getFakeRegExpContext(regex)

        visitParents(context.patternAst, (parent) => {
            for (const dir of directions) {
                const classes = getDeterminismEqClasses(
                    parent.alternatives,
                    dir,
                    context,
                )

                chai.assert.equal(
                    classes.reduce((p, c) => p + c.length, 0),
                    parent.alternatives.length,
                    "expected the number of returned alternatives to be the same as the number of input alternatives.",
                )

                // eslint-disable-next-line @typescript-eslint/require-array-sort-compare -- x
                result[dir][parent.raw] = classes
                    .map((eq) =>
                        [...eq]
                            .sort((a, b) => a.start - b.start)
                            .map((a) => a.raw)
                            .join("|"),
                    )
                    .sort()
            }
        })

        return result
    }

    for (const regex of directionIndependentRegexes) {
        it(regex.toString(), function () {
            const actual = getAllEqClasses(regex)
            chai.assert.deepEqual(
                actual.ltr,
                actual.unknown,
                "expected ltr to be equal to unknown direction",
            )
            chai.assert.deepEqual(
                actual.rtl,
                actual.unknown,
                "expected rtl to be equal to unknown direction",
            )
            chai.expect(actual.ltr).toMatchSnapshot()
        })
    }

    for (const regex of directionalRegexes) {
        it(regex.toString(), function () {
            const actual = getAllEqClasses(regex)
            chai.expect(actual).toMatchSnapshot()
        })
    }
})

function visitParents(
    root: Node,
    onParent: (parent: Alternative["parent"]) => void,
): void {
    visitRegExpAST(root, {
        onAssertionEnter(node) {
            if (node.kind === "lookahead" || node.kind === "lookbehind") {
                onParent(node)
            }
        },
        onCapturingGroupEnter: onParent,
        onGroupEnter: onParent,
        onPatternEnter: onParent,
    })
}

function getFakeRegExpContext(regex: RegExp): RegExpContext {
    const literal = new RegExpParser().parseLiteral(regex.toString())

    const toCharSet: ToCharSet = (node, flags) => {
        return RAA.toCharSet(node, flags ?? literal.flags)
    }

    const partialContext: Partial<RegExpContext> = {
        pattern: regex.source,
        patternAst: literal.pattern,
        flags: literal.flags,
        flagsString: regex.flags,

        toCharSet,
    }

    return partialContext as RegExpContext
}
