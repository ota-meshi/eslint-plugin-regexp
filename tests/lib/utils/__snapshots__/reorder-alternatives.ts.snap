// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDeterminismEqClasses /(?:script|source)_foo|sample/ 1`] = `
Object {
  "(?:script|source)": Array [
    "script",
    "source",
  ],
  "(?:script|source)_foo|sample": Array [
    "(?:script|source)_foo",
    "sample",
  ],
}
`;

exports[`getDeterminismEqClasses /(int|integer)\\b/ 1`] = `
Object {
  "(int|integer)": Array [
    "int",
    "integer",
  ],
  "(int|integer)\\\\b": Array [
    "(int|integer)\\\\b",
  ],
}
`;

exports[`getDeterminismEqClasses />>=?|<<=?|->|--|\\+\\+|&&|\\|\\||[?:~]|<=>|[-+*/%&|^!=<>]=?|\\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\\b/ 1`] = `
Object {
  "ltr": Object {
    "(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)": Array [
      "and",
      "and_eq",
      "bitand",
      "bitor",
      "not",
      "not_eq",
      "or",
      "or_eq",
      "xor",
      "xor_eq",
    ],
    ">>=?|<<=?|->|--|\\\\+\\\\+|&&|\\\\|\\\\||[?:~]|<=>|[-+*/%&|^!=<>]=?|\\\\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\\\\b": Array [
      ">>=?|<<=?|->|--|\\\\+\\\\+|&&|\\\\|\\\\||<=>|[-+*/%&|^!=<>]=?",
      "[?:~]",
      "\\\\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\\\\b",
    ],
  },
  "rtl": Object {
    "(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)": Array [
      "and",
      "and_eq",
      "bitand",
      "bitor",
      "not",
      "not_eq",
      "or",
      "or_eq",
      "xor",
      "xor_eq",
    ],
    ">>=?|<<=?|->|--|\\\\+\\\\+|&&|\\\\|\\\\||[?:~]|<=>|[-+*/%&|^!=<>]=?|\\\\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\\\\b": Array [
      ">>=?|<<=?|->|--|\\\\+\\\\+|&&|\\\\|\\\\||<=>|[-+*/%&|^!=<>]=?",
      "[?:~]",
      "\\\\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\\\\b",
    ],
  },
  "unknown": Object {
    "(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)": Array [
      "and",
      "and_eq",
      "bitand",
      "bitor",
      "not",
      "not_eq",
      "or",
      "or_eq",
      "xor",
      "xor_eq",
    ],
    ">>=?|<<=?|->|--|\\\\+\\\\+|&&|\\\\|\\\\||[?:~]|<=>|[-+*/%&|^!=<>]=?|\\\\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\\\\b": Array [
      ">>=?|<<=?|->|--|\\\\+\\\\+|&&|\\\\|\\\\||<=>|[-+*/%&|^!=<>]=?",
      "[?:~]",
      "\\\\b(?:and|and_eq|bitand|bitor|not|not_eq|or|or_eq|xor|xor_eq)\\\\b",
    ],
  },
}
`;

exports[`getDeterminismEqClasses /\\bcircle|ellipse|closest|farthest|contain|cover\\b/ 1`] = `
Object {
  "\\\\bcircle|ellipse|closest|farthest|contain|cover\\\\b": Array [
    "\\\\bcircle",
    "closest",
    "contain",
    "cover\\\\b",
    "ellipse",
    "farthest",
  ],
}
`;

exports[`getDeterminismEqClasses /\\p{L}|[a-z]|\\d+/u 1`] = `
Object {
  "\\\\p{L}|[a-z]|\\\\d+": Array [
    "\\\\d+",
    "\\\\p{L}|[a-z]",
  ],
}
`;

exports[`getDeterminismEqClasses /a.*b|b.*b|b.*a/ 1`] = `
Object {
  "ltr": Object {
    "a.*b|b.*b|b.*a": Array [
      "a.*b",
      "b.*b|b.*a",
    ],
  },
  "rtl": Object {
    "a.*b|b.*b|b.*a": Array [
      "a.*b|b.*b",
      "b.*a",
    ],
  },
  "unknown": Object {
    "a.*b|b.*b|b.*a": Array [
      "a.*b|b.*b|b.*a",
    ],
  },
}
`;

exports[`getDeterminismEqClasses /a.*b|b.*b|c.*a/ 1`] = `
Object {
  "ltr": Object {
    "a.*b|b.*b|c.*a": Array [
      "a.*b",
      "b.*b",
      "c.*a",
    ],
  },
  "rtl": Object {
    "a.*b|b.*b|c.*a": Array [
      "a.*b|b.*b",
      "c.*a",
    ],
  },
  "unknown": Object {
    "a.*b|b.*b|c.*a": Array [
      "a.*b|b.*b",
      "c.*a",
    ],
  },
}
`;

exports[`getDeterminismEqClasses /a{20}|b{20}|[^]{20}/ 1`] = `
Object {
  "a{20}|b{20}|[^]{20}": Array [
    "a{20}|b{20}|[^]{20}",
  ],
}
`;

exports[`getDeterminismEqClasses /a{20}c|b{20}a|[^]{19}b/ 1`] = `
Object {
  "ltr": Object {
    "a{20}c|b{20}a|[^]{19}b": Array [
      "a{20}c",
      "b{20}a|[^]{19}b",
    ],
  },
  "rtl": Object {
    "a{20}c|b{20}a|[^]{19}b": Array [
      "[^]{19}b",
      "a{20}c",
      "b{20}a",
    ],
  },
  "unknown": Object {
    "a{20}c|b{20}a|[^]{19}b": Array [
      "a{20}c",
      "b{20}a|[^]{19}b",
    ],
  },
}
`;

exports[`getDeterminismEqClasses /a{20}c|b{20}a|[^]{20}b/ 1`] = `
Object {
  "a{20}c|b{20}a|[^]{20}b": Array [
    "[^]{20}b",
    "a{20}c",
    "b{20}a",
  ],
}
`;

exports[`getDeterminismEqClasses /a|aa|aaa|b|bb|bbb/ 1`] = `
Object {
  "a|aa|aaa|b|bb|bbb": Array [
    "a|aa|aaa",
    "b|bb|bbb",
  ],
}
`;

exports[`getDeterminismEqClasses /a|b|c|d/ 1`] = `
Object {
  "a|b|c|d": Array [
    "a",
    "b",
    "c",
    "d",
  ],
}
`;

exports[`getDeterminismEqClasses /a|b|c|d|[a-c]/ 1`] = `
Object {
  "a|b|c|d|[a-c]": Array [
    "a|b|c|[a-c]",
    "d",
  ],
}
`;

exports[`getDeterminismEqClasses /aa\\d+|ba\\d+/ 1`] = `
Object {
  "ltr": Object {
    "aa\\\\d+|ba\\\\d+": Array [
      "aa\\\\d+",
      "ba\\\\d+",
    ],
  },
  "rtl": Object {
    "aa\\\\d+|ba\\\\d+": Array [
      "aa\\\\d+|ba\\\\d+",
    ],
  },
  "unknown": Object {
    "aa\\\\d+|ba\\\\d+": Array [
      "aa\\\\d+|ba\\\\d+",
    ],
  },
}
`;

exports[`getDeterminismEqClasses /aaaaaaaaaaaaaaaaaaaa|bbbbbbbbbbbbbbbbbbbb|[^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^]/ 1`] = `
Object {
  "aaaaaaaaaaaaaaaaaaaa|bbbbbbbbbbbbbbbbbbbb|[^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^]": Array [
    "aaaaaaaaaaaaaaaaaaaa|bbbbbbbbbbbbbbbbbbbb|[^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^][^]",
  ],
}
`;

exports[`getDeterminismEqClasses /ab|bc|ca/ 1`] = `
Object {
  "ab|bc|ca": Array [
    "ab",
    "bc",
    "ca",
  ],
}
`;

exports[`getDeterminismEqClasses /abc/ 1`] = `
Object {
  "abc": Array [
    "abc",
  ],
}
`;

exports[`getDeterminismEqClasses /abc|a/ 1`] = `
Object {
  "ltr": Object {
    "abc|a": Array [
      "abc|a",
    ],
  },
  "rtl": Object {
    "abc|a": Array [
      "a",
      "abc",
    ],
  },
  "unknown": Object {
    "abc|a": Array [
      "abc|a",
    ],
  },
}
`;

exports[`getDeterminismEqClasses /anchor_1_x|anchor_1_y|anchor_2_x|anchor_2_y|reaction_force_x|reaction_force_y|reaction_torque|motor_speed|angle|motor_torque|max_motor_torque|translation|speed|motor_force|max_motor_force|length_1|length_2|damping_ratio|frequency|lower_angle_limit|upper_angle_limit|angle_limits|max_length|max_torque|max_force/ 1`] = `
Object {
  "ltr": Object {
    "anchor_1_x|anchor_1_y|anchor_2_x|anchor_2_y|reaction_force_x|reaction_force_y|reaction_torque|motor_speed|angle|motor_torque|max_motor_torque|translation|speed|motor_force|max_motor_force|length_1|length_2|damping_ratio|frequency|lower_angle_limit|upper_angle_limit|angle_limits|max_length|max_torque|max_force": Array [
      "anchor_1_x",
      "anchor_1_y",
      "anchor_2_x",
      "anchor_2_y",
      "angle|angle_limits",
      "damping_ratio",
      "frequency",
      "length_1",
      "length_2",
      "lower_angle_limit",
      "max_force",
      "max_length",
      "max_motor_force",
      "max_motor_torque",
      "max_torque",
      "motor_force",
      "motor_speed",
      "motor_torque",
      "reaction_force_x",
      "reaction_force_y",
      "reaction_torque",
      "speed",
      "translation",
      "upper_angle_limit",
    ],
  },
  "rtl": Object {
    "anchor_1_x|anchor_1_y|anchor_2_x|anchor_2_y|reaction_force_x|reaction_force_y|reaction_torque|motor_speed|angle|motor_torque|max_motor_torque|translation|speed|motor_force|max_motor_force|length_1|length_2|damping_ratio|frequency|lower_angle_limit|upper_angle_limit|angle_limits|max_length|max_torque|max_force": Array [
      "anchor_1_x",
      "anchor_1_y",
      "anchor_2_x",
      "anchor_2_y",
      "angle",
      "angle_limits",
      "damping_ratio",
      "frequency",
      "length_1",
      "length_2",
      "lower_angle_limit",
      "max_force",
      "max_length",
      "max_torque",
      "motor_force|max_motor_force",
      "motor_speed|speed",
      "motor_torque|max_motor_torque",
      "reaction_force_x",
      "reaction_force_y",
      "reaction_torque",
      "translation",
      "upper_angle_limit",
    ],
  },
  "unknown": Object {
    "anchor_1_x|anchor_1_y|anchor_2_x|anchor_2_y|reaction_force_x|reaction_force_y|reaction_torque|motor_speed|angle|motor_torque|max_motor_torque|translation|speed|motor_force|max_motor_force|length_1|length_2|damping_ratio|frequency|lower_angle_limit|upper_angle_limit|angle_limits|max_length|max_torque|max_force": Array [
      "anchor_1_x",
      "anchor_1_y",
      "anchor_2_x",
      "anchor_2_y",
      "angle|angle_limits",
      "damping_ratio",
      "frequency",
      "length_1",
      "length_2",
      "lower_angle_limit",
      "max_force",
      "max_length",
      "max_torque",
      "motor_force|max_motor_force",
      "motor_speed|speed",
      "motor_torque|max_motor_torque",
      "reaction_force_x",
      "reaction_force_y",
      "reaction_torque",
      "translation",
      "upper_angle_limit",
    ],
  },
}
`;

exports[`getDeterminismEqClasses /device_ios_(?:ipad|ipad_retina|iphone|iphone5|iphone6|iphone6plus|iphone_retina|unknown)\\b/ 1`] = `
Object {
  "(?:ipad|ipad_retina|iphone|iphone5|iphone6|iphone6plus|iphone_retina|unknown)": Array [
    "ipad",
    "ipad_retina",
    "iphone",
    "iphone5",
    "iphone6",
    "iphone6plus",
    "iphone_retina",
    "unknown",
  ],
  "device_ios_(?:ipad|ipad_retina|iphone|iphone5|iphone6|iphone6plus|iphone_retina|unknown)\\\\b": Array [
    "device_ios_(?:ipad|ipad_retina|iphone|iphone5|iphone6|iphone6plus|iphone_retina|unknown)\\\\b",
  ],
}
`;

exports[`getDeterminismEqClasses /int|integer/ 1`] = `
Object {
  "ltr": Object {
    "int|integer": Array [
      "int|integer",
    ],
  },
  "rtl": Object {
    "int|integer": Array [
      "int",
      "integer",
    ],
  },
  "unknown": Object {
    "int|integer": Array [
      "int|integer",
    ],
  },
}
`;
